### PR TITLE
Experimental: Change default behavior to host_only for nvcc_wrapper

### PR DESCRIPTION
**Pull request by jjwilke**
This is only a proposal so people would know what this looks like.  There has been a concern that projects using C++17 or other things `nvcc` doesn't support have trouble since they are forced to use `nvcc_wrapper`.  It potentially makes more sense for targets NOT depending on Kokkos to default to host and projects that do depend on Kokkos to force `nvcc_wrapper` into device mode.

### CODE REVIEW

**Comment by jjwilke** [comment]
So another use case here is Spack. `nvcc_wrapper` is a nightmare to use with Spack right now.  It would actually be nice if we could just manually override Spack to tell it, e.g. `gcc@7.2.0` is actually `nvcc_wrapper` and have it be the host compiler.

**Comment by jjwilke** [comment]
Okay. After more thinking, it will be pretty annoying and challenging to maintain two scripts (nvcc_wrapper and, e.g. kokkos_wrapper) where nvcc_wrapper now calls kokkos_wrapper. The two scripts would have to know about each other's locations (which can be done by converting them to .in files and hardcoding the prefix). I'm starting to prefer just changing the default behavior of nvcc_wrapper. I honestly don't think anyone would notice because a `--device` flag will be silently added everywhere due to `KOKKOS_CXXFLAGS`

**Comment by jjwilke** [comment]
@dalg24 I think this is actually ready for initial review

File: bin/nvcc_wrapper [hunk 1]
```diff
@@ -194,28 +200,48 @@ do
...
+      shared_args="$shared_args $std_flag"
+    else
+      #nope, not okay, we are in nvcc mode
+      bad_std_flag=$1
+      bad_nvcc_std=1
+      if [ -n "$std_flag" ]; then
+         warn_std_flag
+         shared_args=${shared_args/ $std_flag/}
+      fi
+      std_flag=$1
```
Comment (crtrott) [inline bin/nvcc_wrapper]: Doesn't this change the behavior where we earlier used fallback_std_flag?

File: bin/nvcc_wrapper [hunk 2]
```diff
@@ -338,6 +364,10 @@ do
   shift
 done
 
+if [ $host_only -eq 0 -a $bad_nvcc_std -eq 1 ]; then
+  echo "nvcc_wrapper cannot compile for device with unsupported standard $bad_std_flag"
+fi
```
Comment (crtrott) [inline bin/nvcc_wrapper]: ah we now just error out?

File: cmake/kokkos_compiler_id.cmake [hunk 3]
```diff
@@ -5,23 +5,19 @@ SET(KOKKOS_CXX_COMPILER_ID ${CMAKE_CXX_COMPILER_ID})
...
+  STRING(REGEX MATCH "nvcc_wrapper" HAVE_NVCC_WRAPPER ${KOKKOS_COMPILER_EXE})
+  IF (NOT HAVE_NVCC_WRAPPER)
+    MESSAGE(FATAL_ERROR "Using NVCC compiler directly is not supported - must use nvcc_wrapper")
+  ENDIF()
   # SET the compiler id to nvcc.  We use the value used by CMake 3.8.
   SET(KOKKOS_CXX_COMPILER_ID NVIDIA CACHE STRING INTERNAL FORCE)
 
   # SET nvcc's compiler version.
-  EXECUTE_PROCESS(COMMAND ${CMAKE_CXX_COMPILER} --version
+  EXECUTE_PROCESS(COMMAND ${CMAKE_CXX_COMPILER} --version --device
```
Comment (crtrott) [inline cmake/kokkos_compiler_id.cmake]: why not just do this if we have HAVE_NVCC_WRAPPER and let people try use nvcc directly if they really want to?

File: bin/nvcc_wrapper (hunk 2 above)
Comment (jjwilke) [reply bin/nvcc_wrapper]: Yes. CMake tests will run in "host mode" which means there is no longer any scenario where we need to accept bad flags.

File: bin/nvcc_wrapper (hunk 1 above)
Comment (jjwilke) [reply bin/nvcc_wrapper]: Correct. Same as comment below. CMake tests now run in host mode so we don't need to do any fallback flag hacking.

File: cmake/kokkos_compiler_id.cmake (hunk 3 above)
Comment (jjwilke) [reply cmake/kokkos_compiler_id.cmake]: Yeah, we could do that. I hadn't considered the possibility of anyone using straight NVCC.

**Comment by crtrott** [comment]
Doing the RuleLaunch thing instead #3136

**Pull request closed without merging**